Lib/cu2qu/__init__.py

```diff
@@ -16,154 +16,182 @@
-    """Return the point on a cubic bezier curve at time t."""
 
-    (x1, y1), (x2, y2), (x3, y3), (x4, y4) = p
+def _cubic_farthest_fit(pt1,pt2,pt3,pt4,tolerance):
+    """Returns True if the cubic Bezier p entirely lies within a distance
+    tolerance of origin, False otherwise."""
 
-    _t = 1-t
-    t2 = t*t
-    _t2 = _t*_t
-    t3 = t*t2
-    _t3 = _t*_t2
-    _3_t2_t = 3*t2*_t
-    _3_t_t2 = 3*t*_t2
+    if abs(pt2) <= tolerance and abs(pt3) <= tolerance:
+        return True
 
-    return (_t3*x1 + _3_t_t2*x2 + _3_t2_t*x3 + t3*x4,
-            _t3*y1 + _3_t_t2*y2 + _3_t2_t*y3 + t3*y4)
+    # Split.
+    mid = (pt1+3*(pt2+pt3)+pt4)*.125
+    if abs(mid) > tolerance:
+        return False
+    deriv3 = (pt4+pt3-pt2-pt1)*.125
+    return (_cubic_farthest_fit(pt1, (pt1+pt2)*.5, mid-deriv3, mid,tolerance) and
+            _cubic_farthest_fit(mid, mid+deriv3, (pt3+pt4)*.5, pt4,tolerance))
 
+def cubic_farthest_fit(pt1,pt2,pt3,pt4,tolerance):
+    """Returns True if the cubic Bezier p entirely lies within a distance
+    tolerance of origin, False otherwise."""
 
-def cubic_approx(p, t):
-    """Approximate a cubic bezier curve with a quadratic one."""
+    if abs(pt1) > tolerance or abs(pt4) > tolerance:
+        return False
 
-    p1 = lerp_pt(p[0], p[1], 1.5)
-    p2 = lerp_pt(p[3], p[2], 1.5)
-    return p[0], lerp_pt(p1, p2, t), p[3]
+    if abs(pt2) <= tolerance and abs(pt3) <= tolerance:
+        return True
 
+    # Split.
+    mid = (pt1+3*(pt2+pt3)+pt4)*.125
+    if abs(mid) > tolerance:
+        return False
+    deriv3 = (pt4+pt3-pt2-pt1)*.125
+    return (_cubic_farthest_fit(pt1, (pt1+pt2)*.5, mid-deriv3, mid,tolerance) and
+            _cubic_farthest_fit(mid, mid+deriv3, (pt3+pt4)*.5, pt4,tolerance))
 
-def calc_intersect(p):
-    """Calculate the intersection of ab and cd, given [a, b, c, d]."""
 
-    a, b, c, d = p
-    ab = vector(a, b)
-    cd = vector(c, d)
-    p = -ab[1], ab[0]
-    try:
-        h = dot(p, vector(c, a)) / dot(p, cd)
-    except ZeroDivisionError:
-        raise ValueError('Parallel vectors given to calc_intersect.')
-    return translate(c, scale(cd, h))
-
-
-def cubic_approx_spline(p, n):
+def cubic_approx_spline(p, n, tolerance):
     """Approximate a cubic bezier curve with a spline of n quadratics.
 
     Returns None if n is 1 and the cubic's control vectors are parallel, since
     no quadratic exists with this cubic's tangents.
     """
 
     if n == 1:
-        try:
-            p1 = calc_intersect(p)
-        except ValueError:
+        qp1 = calc_intersect(*p)
+        if qp1 is None:
+            return None
+        p0 = p[0]
+        p3 = p[3]
+        p1 = p0+(qp1-p0)*(2./3)
+        p2 =  p3+(qp1-p3)*(2./3)
+        if not cubic_farthest_fit(0,
+                                  p1-p[1],
+                                  p2-p[2],
+                                  0, tolerance):
             return None
-        return [p[0], p1, p[3]]
+        return p0, p1, p2, p3
```

Comment:
Sorry yea, I meant "length 3" as in three points, so a spline with 1 segment.

Other things look good except updating tests, which I'm about to do.